server/request.go

```diff
@@ -0,0 +1,206 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/asdine/storm"
+	"github.com/go-chi/chi"
+	"github.com/kochman/hotshots/log"
+)
+
+/*
+ * Response Structs
+ */
+
+type GetPhotoIDsResponse struct {
+	Success bool
+	IDs     []string
+}
+
+type GetPhotosResponse struct {
+	Success bool
+	Photos  []Photo
+}
+
+type PostPhotoResponse struct {
+	Success bool
+	NewID   string `json:",omitempty"`
+}
+
+type GetPhotoMetadataResponse struct {
+	Success bool
+	Photo   Photo
+}
+
+type ErrorResponse struct {
+	Success bool
+	Error   string
+}
+
+type NotFoundResponse struct {
+	Success bool
+	Error   string
+	URI     string
+}
+
+/*
+ * Handlers
+ */
+
+func NotFound(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(404)
+	v := NotFoundResponse{
+		Success: false,
+		Error:   "Requested URI not found",
+		URI:     r.RequestURI,
+	}
+	WriteJsonResponse(v, w)
+}
+
+func (s *Server) PhotoCtx(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		photoID := chi.URLParam(r, "pid")
+		var photo Photo
+		if err := s.db.One("ID", photoID, &photo); err != nil {
+			log.Info(err)
+			WriteError("Unable to find photo id", 404, w)
+			return
+		}
+		ctx := context.WithValue(r.Context(), "photo", photo)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+func (s *Server) GetPhotos(w http.ResponseWriter, r *http.Request) {
+	var photos []Photo
+	if err := s.db.All(&photos); err != nil {
+		log.Error(err)
+		WriteError("Unable to query photos", 500, w)
+		return
+	}
+
+	v := GetPhotosResponse{
+		Success: true,
+		Photos:  photos,
+	}
+	WriteJsonResponse(v, w)
+}
+
+func (s *Server) GetPhotoIDs(w http.ResponseWriter, r *http.Request) {
+	var photos []Photo
+	if err := s.db.All(&photos); err != nil {
+		log.Error(err)
+		WriteError("Unable to query photos", 500, w)
+		return
+	}
+	ids := make([]string, len(photos))
+	for i, photo := range photos {
+		ids[i] = photo.ID
+	}
+
+	v := GetPhotoIDsResponse{
+		Success: true,
+		IDs:     ids,
+	}
+	WriteJsonResponse(v, w)
+}
+
+func (s *Server) PostPhoto(w http.ResponseWriter, r *http.Request) {
+	r.ParseMultipartForm(1 << 23) // 8M max memory
```

Comment:
Should we up this a bit? Anything up to maybe 20–30 MB would be a reasonable size for an uploaded JPEG, and if we up this a little bit (maybe 32 MB? Go sets its `defaultMaxMemory` to that) then we can avoid temp files touching the disk at all. We're not RAM-constrained on the server.

Reply:
8mb is Max ram it uses, anything over will go to file system tmp space, up to you